test/extended/networking/acl_audit_log.go

```diff
@@ -0,0 +1,188 @@
+	InOVNKubernetesContext(
+		func() {
+			f := oc.KubeFramework()
+			It("should ensure acl logs are created and correct", func() {
+				makeNamespaceACLLoggingEnabled(oc, f.Namespace)
+				testACLLogging(f, oc)
+			})
+		},
+	)
+})
+
+func makeNamespaceACLLoggingEnabled(oc *exutil.CLI, ns *kapiv1.Namespace) {
+	By("setting the k8s.ovn.org/acl-logging annotation for the namespace: " + oc.KubeFramework().Namespace.Name)
+	var err error
+	if ns.Annotations == nil {
+		ns.Annotations = make(map[string]string, 1)
+	}
+	ns.Annotations["k8s.ovn.org/acl-logging"] = `{ "deny": "alert", "allow": "alert" }`
+	_, err = oc.AdminKubeClient().CoreV1().Namespaces().Update(context.Background(), ns, metav1.UpdateOptions{})
+	expectNoError(err)
+}
+
+// Test the Network policy audit logging feature
+func testACLLogging(f *e2e.Framework, oc *exutil.CLI) {
+	// We launch 3 pods total; pod[0] and pod[1] will end up on node[0] in ns "acl-logging-on" , and pod[2]
+	// will end up on node[1] in ns "acl-logging off", to know which acl-logging container to look in
+	var nodes [2]*kapiv1.Node
+	var ns [2]string
+	var err error
+
+	ns[0] = f.Namespace.Name
+
+	makeNamespaceScheduleToAllNodes(f)
+	nsNoACLLog := oc.SetupNamespace()
+	expectNoError(err)
+
+	By("making namespace " + nsNoACLLog + " with acl-logging disabled")
+
+	ns[1] = nsNoACLLog
+
+	nodes[0], nodes[1], err = findAppropriateNodes(f, DIFFERENT_NODE)
+	if err != nil {
+		expectNoError(err)
+	}
+
+	var pod, ip [3]string
+	var errs [3]error
+	By("making two pods in the acl-logging-on namespace and one in the acl-logging-off namespace")
+	// make the first two pods in ns acl-logging-on on node[0] and the other in acl-logging-off on node[1]
+	for i := range pod {
+		pod[i] = fmt.Sprintf("acl-logging-%d", i+1)
+
+		testPod := e2epod.NewAgnhostPod(ns[i/2], pod[i], nil, nil, nil, "netexec", "--http-port=80", "--udp-port=90")
+		testPod.Spec.NodeName = nodes[i/2].Name
+		_, err := f.ClientSet.CoreV1().Pods(ns[i/2]).Create(context.TODO(), testPod, metav1.CreateOptions{})
+		expectNoError(err)
+	}
+	// make sure pods come up
+	for i := range pod {
+		ip[i], errs[i] = waitForACLLoggingPod(f, ns[i/2], pod[i])
+		expectNoError(errs[i])
+	}
+	By("making \"default deny\" and \"allow from same namespace\" network policies")
+	// make two network policies, a default deny all and allow from same namespace in ns "acl-logging-on"
+	_, err = makeDenyAllPolicy(f, ns[0])
+	expectNoError(err)
+
+	_, err = makeFromSameNSPolicy(f, ns[0])
+	expectNoError(err)
+	By("sending allowed traffic between pods in same namespace")
+	// Ping pod[0] from pod[1] which should succeed since it's in the same namespace with out == ip[0]
+	// Should hit the `allow-same-namespace` networkpoliocy
+	out, _, err := oc.AsAdmin().Run("exec").Args(pod[1], "-n", ns[0], "--", "curl", "-m", "1", ip[0]+"/clientip").Outputs()
+	if !strings.Contains(out, ip[1]) || err != nil {
+		expectNoError(err)
```

Comment:
if the first conditional is true here, this will proceed since `err` is nil.

Reply:
Got it